src/Idris/Parser/Expr.hs

```diff
@@ -1288,16 +1290,26 @@ do_ :: SyntaxInfo -> IdrisParser PDo
 do_ syn
      = P.try (do keyword "let"
                  (i, ifc) <- withExtent name
-                 ty <- P.option Placeholder (do lchar ':'
-                                                expr' syn)
+                 ty' <- P.optional (do lchar ':'
+                                       expr' syn)
                  reservedOp "="
-                 (e, fc) <- withExtent $ expr syn
-                 return (DoLet fc RigW i ifc ty e))
+                 (e, fc) <- withExtent $ expr (syn { withAppAllowed = False })
+                 -- If there is an explicit type, this can’t be a pattern-matching let, so do not parse alternatives
+                 case ty' of
+                   Just ty -> return (DoLet fc RigW i ifc ty e)
+                   Nothing ->
+                     P.option (DoLet fc RigW i ifc Placeholder e)
+                              (do lchar '|'
+                                  ts <- P.sepBy1 (do_alt (syn { withAppAllowed = False })) (lchar '|')
```

Comment:
If there's a reason not to support the above case, then putting `when (isJust ty') $ fail "<reason>"` above this line instead of using the `case` will explain to the user what happened.